libindy/src/services/wallet/encryption.rs

```diff
@@ -114,22 +129,50 @@ pub(super) fn decrypt_tags(etags: &Option<Vec<Tag>>, tag_name_key: &[u8], tag_va
     }
 }
 
+pub(super) fn decrypt_storage_record(record: &StorageEntity, keys: &Keys) -> Result<WalletRecord, WalletError> {
+    let decrypted_name = decrypt_merged(&record.name, &keys.name_key)?;
+    let decrypted_name = String::from_utf8(decrypted_name)?;
+
+    let decrypted_value = match record.value {
+        Some(ref value) => {
+            let decrypted_value_key = decrypt_merged(&value.key, &keys.value_key)?;
+            let decrypted_value = decrypt_merged(&value.data, &decrypted_value_key)?;
+            Some(String::from_utf8(decrypted_value)?)
+        },
+        None => None
+    };
+
+    let decrypted_type = match record.type_ {
+        Some(ref type_) => {
+            let decrypted_type = decrypt_merged(type_, &keys.type_key)?;
+            Some(String::from_utf8(decrypted_type)?)
+        },
+        None => None,
+    };
+
+    let decrypted_tags = decrypt_tags(&record.tags, &keys.tag_name_key, &keys.tag_value_key)?;
+    Ok(WalletRecord::new(decrypted_name, decrypted_type, decrypted_value, decrypted_tags))
+}
+
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
     #[test]
     fn test_encrypt_decrypt_tags() {
-        let tags = r#"{"tag1":"value1", "tag2":"value2", "~tag3":"value3"}"#;
+        let mut tags = HashMap::new();
+        tags.insert("tag1".to_string(), "value1".to_string());
+        tags.insert("tag2".to_string(), "value2".to_string());
+        tags.insert("~tag3".to_string(), "value3".to_string());
 
```

Comment:
I suggest using the sequence of initialization string and then parsing to HashMap instead of building HashMap because it is much more readable and similar to actual usage